edx/edx-ora2
# [BD-05] [TNL-7310] [BB-3388] Pluggable ORA Submission Editor

***This PR has been split into two separate PRs -***
- https://github.com/edx/edx-ora2/pull/1574 - Implementation of Pluggable Architecture based on RequireJS
- https://github.com/edx/edx-ora2/pull/1576 - Adds support for TinyMCE as WYSIWYG Editor

----
This PR adds pluggable submission editor support for ORA.

In order to add multiple submission editor along with pluggability following flow has been implemented -

1. ORA gets available editor configurations from ``editor_config.py`` where it's defined as -
```python
AVAILABLE_EDITORS = {
    'text': {  # key is the identification for the editor
        'display': 'Simple Text Editor',  # displays name in studio settings selectbox
        'js': ['/static/js/openassessment-editor-textarea.js'],  # this JS file tells ORA how this editor behaves 
        'css': []  # this CSS files will be included
    },
    'tinymce': {
        'id': 'tinymce',
        'display': 'WYSIWYG Editor',
        'js': ['/static/js/openassessment-editor-tinymce.js'],
        'css': ['/static/js/vendor/tinymce/js/tinymce/skins/studio-tmce4/skin.min.css'] 
    },
}
```
2. This ``AVAILABLE_EDITORS`` can be overridden by the ``ORA_AVAILABLE_EDITORS`` settings.
3. ORA uses the ``ResponseEditorLoader`` to load the selected editor. Selected editor and Available editor configuration are passed via JS context.
4. ``ResponseEditorLoader`` uses ``RequireJS`` to load JS files from the editor config.
4. The file that tells how a ``Editor`` behaves should have the following statucture -
```javascript
(function (define) {
  define(['additional js for this editor to work'], () => {
    class Editor {
      load(elements) {} // loads the editor

      setOnChangeListener(handler) {} // binds handler to editor's on change
      
      response(texts) {} // sets editor value if texts are present else returns existing values
    }  

    // return a function, to be able to create a new instance every time. 
    // otherwise, it can collide with other ORA block in same subsection.
    return function () {
      return new EditorTinymce();
    };
  });

// Studio has define directly on the window, LMS has RequireJS.
}).call(window, window.define || window.RequireJS.define);

```
5. Both ``js/openassessment-editor-textarea.js`` and ``js/openassessment-editor-tinymce.js`` follows this structure. 

**JIRA tickets**: 
- https://openedx.atlassian.net/browse/TNL-7310
- https://tasks.opencraft.com/browse/BB-3388

~~**Discussions**:~~

**Dependencies**: None

**Screenshots**: 
Studio:
![Screenshot from 2020-12-23 12-46-12](https://user-images.githubusercontent.com/1010244/103025070-42445300-457b-11eb-8f94-6112d6e3c8bb.png)

TinyMCE on LMS:
![Screenshot from 2020-12-23 23-12-58](https://user-images.githubusercontent.com/1010244/103025128-5c7e3100-457b-11eb-9663-30c67f2714aa.png)

TinyMCE on read-only mode:
![Screenshot from 2020-12-23 23-13-36](https://user-images.githubusercontent.com/1010244/103025170-6f910100-457b-11eb-9f50-80b1768c2533.png)


**Sandbox URL**: TBD

**Merge deadline**: None

**Testing instructions**:

1. Clone this PR on your local devstack. Install using ``make install-local-ora``. 
2. Create an ORA block. Select ``WYSIWYG Editor`` from settings according to the 1st screenshot.
3. Publish the ORA block. Click ``View Live``.
4. Check if TinyMCE is loaded on the textarea.
5. Write on the editor, check if on content change Submission & Save buttons become enabled.
6. Click Save. Reload the page, see if save works.
7. Click submit to see if submission works.
8. On the training and other subsequent steps, the demo response should be loaded using TinyMCE as well.
9. To test how an external app can add a new editor, install https://github.com/shimulch/edx-ora2-codemirror-example 

**Author notes and concerns**:

1. This implementation utilizes the availability of ``RequireJS`` on both LMS and Studio. The TinyMCE editor is also dependent on the availability of the plugin on Studio and LMS.
2. This method now doesn't work on Studio. Because the static JS files are not in Studio. We need to use an absolute URL from LMS to make it work. Currently, I am looking for the correct place to build it.
3. This doesn't work on the instructor dashboard as well, looking for a solution for that.
4. Many tests are failing. But first I want to verify if this is the correct way to implement this.
5. One thing I've noticed which might be missed during discovery is we can use multiple editors in a single ORA problem. The way we are selecting editor doesn't accommodate different editors for each prompt. For example - using both codemirror & tinymce for the same ORA problem. To do that we need this editor to select settings on the Prompt tab instead of the settings tab.

**Reviewers**
